packages/core/src/bus/BusInstruction.ts

```diff
@@ -0,0 +1,27 @@
+import { randomBytes } from 'crypto';
+
+export abstract class BusInstruction<T> {
+  public readonly type!: string;
+  public readonly payload!: T;
+  public readonly correlationId!: string;
+  public readonly createdAt!: Date;
+
+  constructor(
+    payload: T,
+    type?: string,
+    correlationId?: string,
+    createdAt?: Date
+  ) {
+    this.payload = payload;
+    this.type = type || this.getType(payload);
+    this.correlationId =
+      correlationId || randomBytes(32).toString('hex').slice(0, 32);
+    this.createdAt = createdAt || new Date();
+  }
+
+  private getType(payload: T) {
+    const { constructor } = Object.getPrototypeOf(payload);
```

Comment:
The type is used to route events and commands to the particular handler. It should be the name of an event or command, it has nothing to do with the payload.

```suggestion
  private getType(): string {
    const { constructor } = Object.getPrototypeOf(this);
```

Reply:
this method was extract to utils